controlplane/kubeadm/internal/workload_cluster_coredns.go

```diff
@@ -293,19 +300,20 @@ func extractImageVersion(tag string) (string, error) {
 // validateCoreDNSImageTag returns error if the versions don't meet requirements.
 // Some of the checks come from
 // https://github.com/coredns/corefile-migration/blob/v1.0.6/migration/migrate.go#L414
-func validateCoreDNSImageTag(fromVersion, toVersion string) error {
-	from, err := semver.Parse(fromVersion)
+func validateCoreDNSImageTag(fromTag, toTag string) error {
+	from, err := util.ParseMajorMinorPatch(fromTag)
 	if err != nil {
-		return errors.Wrapf(err, "failed to semver parse %q", fromVersion)
+		return errors.Wrapf(err, "failed to parse CoreDNS current version %q", fromTag)
 	}
-	to, err := semver.Parse(toVersion)
+	to, err := util.ParseMajorMinorPatch(toTag)
 	if err != nil {
-		return errors.Wrapf(err, "failed to semver parse %q", toVersion)
+		return errors.Wrapf(err, "failed to parse CoreDNS target version %q", toTag)
 	}
-	if from.Compare(to) >= 0 {
-		return fmt.Errorf("toVersion %q must be greater than fromVersion %q", to.String(), from.String())
+	// make sure that the version we're upgrading to is greater than the current one,
+	// or if they're the same version, the raw tags should be different (e.g. allow from `v1.17.4-somevendor.0` to `v1.17.4-somevendor.1`).
+	if x := from.Compare(to); x > 0 || (x == 0 && fromTag == toTag) {
+		return fmt.Errorf("toVersion %q must be greater than fromVersion %q", toTag, fromTag)
```

Comment:
i think this is better. it would still allow:
```
from: v1.17.4-vendor.1
to: v1.17.4-vendor.0
```
and consider the `to` version as newer.
but it's hard to manage that for the time being, without standardization.

/lgtm

Reply:
Thanks! Yeah I thought about that, it's going to be allowed for now until we something better as you said

Reply:
i guess, fallback to string comparison will partially work here too:
```
if x := from.Compare(to); x > 0 || (x == 0 && fromTag >= toTag) {
```